### PR TITLE
Issue #105 - Refactor `MMENSDateWrapper` into `MMEDate`

### PR DESCRIPTION
**Pull request by alfwatt**
Implement #105

### CODE REVIEW

File: MapboxMobileEvents/MMEDate.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#import <Foundation/Foundation.h>
+
+/*!
+@brief MMEDate is a subclas of NSDate which stores information needed to compute clock offsets from servers
+*/
+@interface MMEDate : NSDate <NSSecureCoding>
+/*! @biref store the recorded time offset for the date when created using the +[MMEDate date] or +[MMEDate dateWithOffsetFromServer:(NSTimeInterval)offset]` methods */
+@property(nonatomic,assign,readonly) NSTimeInterval offsetFromServer;
+
+/*! @brief computes, reccords and returns the time offset from the server's time frame */
```
Comment (rclee) [inline MapboxMobileEvents/MMEDate.h]:
```suggestion
/*! @brief computes, records and returns the time offset from the server's time frame */
```

File: MapboxMobileEvents/MMEDate.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+/*! @brief returns the date, with the offsetFromServer adjustment applied, putting the date in the server's time frame  */
+- (NSDate*) offsetToServer;
+
+@end
+
+#pragma mark -
+
+@interface NSDate (MMEDate)
+
+- (NSDate *)mme_oneDayLater; // --> (instanctype) startOfTomorrow;
```
Comment (rclee) [inline MapboxMobileEvents/MMEDate.h]:
```suggestion
- (NSDate *)mme_oneDayLater; // --> (instancetype) startOfTomorrow;
```

**Comment by alfwatt** [comment]
Added tests for MMEDate, deferred the system time work to #107

File: MapboxMobileEventsTests/MMEEventsManagerTests.mm [hunk 3]
```diff
@@ -917,9 +910,6 @@ - (void)pushEvent:(MMEEvent *)event;
         
         beforeEach(^{
             dateString = @"A nice date";
-            spy_on(dateWrapper);
-            dateWrapper stub_method(@selector(formattedDateStringForDate:)).and_return(dateString);
-            eventsManager.dateWrapper = dateWrapper;
```
Comment (rclee) [inline MapboxMobileEventsTests/MMEEventsManagerTests.mm:917]:
```suggestion
            NSDateFormatter *dateFormatter = MMEDate.iso8601DateFormatter;
            spy_on(dateFormatter);
            dateFormatter stub_method(@selector(stringFromDate:)).and_return(dateString);
            
```

File: MapboxMobileEventsTests/MMEEventsManagerTests.mm [hunk 4]
```diff
@@ -788,10 +784,7 @@ - (void)pushEvent:(MMEEvent *)event;
...
-            
+
```
Comment (rclee) [inline MapboxMobileEventsTests/MMEEventsManagerTests.mm]:
```suggestion
            NSDateFormatter *dateFormatter = MMEDate.iso8601DateFormatter;
            spy_on(dateFormatter);
            dateFormatter stub_method(@selector(stringFromDate:)).and_return(dateString);
            
```

Comment (rclee) [reply MapboxMobileEventsTests/MMEEventsManagerTests.mm]: (never used these suggested changes -- so, feel free to do what you want with that)

**Review by rclee — DISMISSED** [review]
I think we're good to go after the tests are fixed 👍

File: MapboxMobileEvents/MMEDate.m [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        _offsetFromServer = [aDecoder decodeDoubleForKey:@"offsetFromServer"];
+    }
+
+    return self;
+}
+
+- (void)encodeWithCoder:(NSCoder *)aCoder {
+    [aCoder encodeDouble:_sinceReferenceDate forKey:@"sinceReferenceDate"];
+    [aCoder encodeDouble:_offsetFromServer forKey:@"offsetFromServer"];
+    [aCoder encodeInteger:1 forKey:@"MMEDateVersion"];
```
Comment (rclee) [inline MapboxMobileEvents/MMEDate.m]: Should we add some constants for all of these keys?

Comment (alfwatt) [reply MapboxMobileEvents/MMEDate.m]: Yes, and there should be a version check

File: MapboxMobileEvents/MMEDate.h (hunk 2 above)
Comment (alfwatt) [reply MapboxMobileEvents/MMEDate.h]: Kill all non-doc comments in headers…

**Comment by alfwatt** [comment]
Tests are passing locally, ready for final review.

File: MapboxMobileEventsTests/MMEDateTests.mm [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+SPEC_BEGIN(MMEDateSpec)
+
+describe(@"MMEDate", ^{
+
+    NSTimeInterval const interval = 60; // just a minute
+
+    context(@"+ recordTimeOffsetFromServer:", ^{
+        it(@"computes offsets from server time", ^{
+            NSDate* serverTime = [NSDate dateWithTimeIntervalSinceNow:interval];
```
Comment (rclee) [inline MapboxMobileEventsTests/MMEDateTests.mm]: nit: several pointer position misplacements in this file

File: MapboxMobileEventsTests/MMEMetricsManager.mm [hunk 7]
```diff
@@ -56,18 +56,17 @@
...
         context(@"when preparing attributes", ^{
-            beforeEach(^{
-                [manager attributes];
-            });
-            
-            it(@"should set dateUTCString with the correct format", ^{
-                [dateFormatter dateFromString:manager.metrics.dateUTCString] should_not be_nil;
+
+            it(@"should set MMEEventDateUTC attributes with the correct format", ^{
+                NSDictionary* attributes = [manager attributes];
```
Comment (rclee) [inline MapboxMobileEventsTests/MMEMetricsManager.mm]: nit: pointer position

File: MapboxMobileEventsTests/MMEDateTests.mm [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+            NSKeyedUnarchiver* unarchiver = [NSKeyedUnarchiver.alloc initForReadingWithData:nowData];
+            unarchiver.requiresSecureCoding = YES;
+            MMEDate* then = [unarchiver decodeObjectOfClass:MMEDate.class forKey:NSKeyedArchiveRootObjectKey];
+            then should_not equal(nil);
+            then.timeIntervalSinceReferenceDate should equal(now.timeIntervalSinceReferenceDate);
+        });
+
+        it(@"should write encoded data to a file", ^{
+            [NSKeyedArchiver archiveRootObject:now toFile:tempFile];
+            [NSFileManager.defaultManager fileExistsAtPath:tempFile] should equal(YES);
```
Comment (rclee) [inline MapboxMobileEventsTests/MMEDateTests.mm]:
```suggestion
            [NSFileManager.defaultManager fileExistsAtPath:tempFile] should be_truthy;
```
Another fun cedary thing

File: MapboxMobileEventsTests/MMEMetricsManager.mm [hunk 9]
```diff
@@ -56,18 +56,17 @@
...
+
+                [dateFormatter dateFromString:attributes[MMEEventDateUTC]] should_not be_nil;
```
Comment (rclee) [inline MapboxMobileEventsTests/MMEMetricsManager.mm]: I should have wrote this before to check the proper format "yyyy-MM-dd"...

**Review by rclee — DISMISSED** [review]
Sorry, for all the nit-picks -- everything else looks solid.

File: MapboxMobileEvents/MMEDate.h [hunk 10]
```diff
@@ -0,0 +1,48 @@
+#import <Foundation/Foundation.h>
+
+/*! @brief MMEDate is a subclass of NSDate which stores information needed to compute clock offsets from servers */
+@interface MMEDate : NSDate
+
+/*! @brief store the recorded time offset for the date when created using the +[MMEDate date] or +[MMEDate dateWithOffsetFromServer:(NSTimeInterval)offset]` methods */
+@property(nonatomic,assign,readonly) NSTimeInterval offsetFromServer;
+
+/*! @brief computes, reccords and returns the time offset from the server's time frame */
```
[the remark this answers is not in the document]
Comment (alfwatt) [reply MapboxMobileEvents/MMEDate.h]: this is why nobody writes documentation ;)

**Review by rclee — DISMISSED** [review]
🚀

**Comment by alfwatt** [comment]
Fixed up the last few test nits as well

**Review by rclee — APPROVED** [review]
Thanks Alf, looks great!

**Pull request merged**